test/state/mpt.cpp

```diff
@@ -132,12 +116,37 @@ class MPTNode
                                     std::move(br);
     }
 
-    /// Finds the position at witch two paths differ.
-    static size_t mismatch(const Path& p1, const Path& p2) noexcept
+    /// The information how two paths are different.
+    struct MismatchResult
     {
-        assert(p1.length <= p2.length);
-        return static_cast<size_t>(
-            std::mismatch(p1.nibbles, p1.nibbles + p1.length, p2.nibbles).first - p1.nibbles);
+        static constexpr uint8_t invalid = 0xff;
+
+        Path common;                ///< Common prefix.
+        uint8_t nibble1 = invalid;  ///< Mismatched nibble in path 1. Invalid if paths match.
+        uint8_t nibble2 = invalid;  ///< Mismatched nibble in second path.
+        Path tail1;                 ///< Remaining tail after the nibble in path 1.
+        Path tail2;                 ///< Remaining tail after the nibble in path 2.
+    };
+
+    /// Finds the difference between two paths.
+    static MismatchResult mismatch(const Path& p1, const Path& p2) noexcept
+    {
+        assert(p1.length <= p2.length);  // no support for inserting keys of different lengths
+        const auto [it1, it2] = std::mismatch(p1.nibbles, p1.nibbles + p1.length, p2.nibbles);
+
+        const Path common{static_cast<size_t>(it1 - p1.nibbles), p1.nibbles};
+        const auto matched = common.length == p1.length;
+        const auto tail_start = common.length + 1;
+        assert(!matched || p1.length != p2.length);  // matched => paths have different lengths
```

Comment:
stronger assertion might also make it easier to understand what's happening 
```suggestion
        assert(!matched || p1.length < p2.length);  // matched => paths have different lengths
```